sensor/common/compliance/nodescan_handler_impl.go

```diff
@@ -0,0 +1,81 @@
+package compliance
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/centralsensor"
+	"github.com/stackrox/rox/pkg/concurrency"
+)
+
+type nodeScanHandlerImpl struct {
+	nodeScans <-chan *storage.NodeScanV2
+	toCentral chan *central.MsgFromSensor
+
+	stopC    concurrency.ErrorSignal
+	stoppedC concurrency.ErrorSignal
+}
+
+func (c *nodeScanHandlerImpl) Capabilities() []centralsensor.SensorCapability {
+	return []centralsensor.SensorCapability{centralsensor.NodeScanningV2}
+}
+
+func (c *nodeScanHandlerImpl) ResponsesC() <-chan *central.MsgFromSensor {
+	return c.toCentral
+}
+
+func (c *nodeScanHandlerImpl) Start() error {
+	go c.run()
+	return nil
+}
+
+func (c *nodeScanHandlerImpl) Stop(err error) {
+	c.stopC.SignalWithError(err)
+}
+
+func (c *nodeScanHandlerImpl) Stopped() concurrency.ReadOnlyErrorSignal {
+	return &c.stoppedC
+}
+
+func (c *nodeScanHandlerImpl) ProcessMessage(msg *central.MsgToSensor) error {
+	// This component doesn't actually process or handle any messages sent from Central to Sensor (yet).
+	// It uses the sensor component so that the lifecycle (start, stop) can be handled when Sensor starts up.
+	return nil
+}
+
+func (c *nodeScanHandlerImpl) run() {
+	defer c.stoppedC.Signal()
+	for {
+		select {
+		case <-c.stopC.Done():
+			c.stoppedC.SignalWithError(c.stopC.Err())
+
```

Comment:
I think I understand that the intent of this code is to arrange a graceful shutdown.
`c.stopC` is when the shutdown was requested. `c.stoppedC` is when the handler actually shut down.

I understand that you likely copied the pattern found in `command_handler_impl.go`, but I think the pattern is wrong.

This specific fragment will just make `for` loop repeat quickly eating up CPU until `c.nodeScans` channel is closed.

I believe the implementation should be fixed here and propagated to other handlers.

I find https://github.com/stackrox/stackrox/blob/master/sensor/common/sensor/central_sender_impl.go to be more clean. Hopefully it can serve as a reference implementation.

A unit test for this handler will be very handy to demonstrate and validate its behavior.

Reply:
> I find https://github.com/stackrox/stackrox/blob/master/sensor/common/sensor/central_sender_impl.go to be more clean. Hopefully it can serve as a reference implementation.

I would love to follow up on this, as I see there several issues that makes the concurrent code difficult to follow. These are not no-goes, but some patterns that I learnt are violated. Examples:
- `forwardResponses` writes to a channel `to` that is passed as an external parameter and never closes the channel to signal that there will no more messages.
- `componentMsgsC` is never closed, so why to have the code that handles the situation when the channel is closed?
- `forwardResponses` stops working as soon as `for !s.stopC.IsDone() {` - this is the same situation as here, we drop messages on the floor instead of forwarding them, as we do not have any guarantee that the stream will still be up. 

I am not saying that the code is not working, but I have several rules that I follow when working with concurrency:
- the goroutine that writes to a channel is responsible for creating the channel and closing it,
- only the writer knows that there will be no more messages sent to a channel. If that happens, the channel should be closed to inform all readers that there will be no more messages, 
- there is exactly one writer goroutine for a channel, but there may be many readers.

Having said that, I am curious to learn why do you think that the code `sensor/common/sensor/central_sender_impl.go` is cleaner.

> This specific fragment will just make for loop repeat quickly eating up CPU until c.nodeScans channel is closed.

Yes, this is an issue and should be addressed. I guess adding `return` should do the job, but I will check that.

> A unit test for this handler will be very handy to demonstrate and validate its behavior.

Agreed! Will try to produce some tests for this.

Reply:
Your notes about channels are likely right. I did not check how channels are handled but I definitely saw that some of them aren't closed.

I brought `central_sender_impl.go` because it more accurately deals with `c.stopC` and `c.stoppedC` than this code. Particularly, there should be only one place that sets `c.stoppedC` - the deferred hook. Everything else like error conditions or closed channels should just initiate the shut down by setting `c.stopC`. Signalling that the shutdown _completed_ (which `c.stoppedC` is supposed to do) in the middle of the function seems wrong to me.

Reply:
This should be addressed in [2077d2a](https://github.com/stackrox/stackrox/pull/3533/commits/2077d2a1e49c6e17960db9da061492231098b338)